backend/src/api/OAuth2Middleware.ts

```diff
@@ -0,0 +1,79 @@
+import { Application, Request, Response, NextFunction } from 'express';
+import { AuthenticateOptions } from 'oauth2-server';
+import { config } from '../config';
+import { ResponseErrorHandler } from './ApiMiddleware';
+import Session, { SessionData } from '../session/Session';
+import DB from '../db/DB';
+import { Logger } from 'winston';
+
+/**
+ * OAuth2 authentication middleware.
+ *
+ * OAuth2 Flow Integration:
+ *   - Checks for the presence of an Authorization header.
+ *   - Invokes the express‑oauth‑server middleware (configured on the app) to authenticate the request.
+ *   - If authentication is successful, the token (containing user information) is available via res.locals.oauth.token.
+ *
+ * Additional Checks:
+ *   - Verifies the token's associated user; for example, it ensures a particular disallowed userId is not allowed.
+ *
+ * Session Integration:
+ *   - When authentication succeeds, the middleware updates session data
+ *     via the Session class.
+ *
+ * @param app Express application instance that has been configured with OAuth2.
+ * @param db Database instance (used for sessions persistence).
+ * @param logger Logger instance for logging errors and warnings.
+ * @returns An Express middleware function that handles OAuth2 authentication.
+ */
+export default function OAuth2Authenticate(app: Application, db: DB, logger: Logger) {
```

Comment:
this needs proper types

Reply:
I think all three are correct, which one is wrong specifically?

Reply:
and its intermediate result needs an explicit type as well